Modules/arraymodule.c

```diff
@@ -2995,6 +3003,70 @@ static PyType_Spec array_spec = {
     .slots = array_slots,
 };
 
+static inline int
+array_subscr_guard(PyObject *lhs, PyObject *rhs)
```

Comment:
What about specializing also on the rhs? (on int or slices)

It would make the code more complex, as the action would have to be split or duplicated. But since performance is the reason for adding this mechanism, maybe the extra specialization is needed.

Reply:
Indeed, we should do that. At the moment I'm just adding the plumbing and implementing a very basic specialisation.